clang/lib/Sema/SemaCast.cpp

```diff
@@ -200,12 +200,17 @@ namespace {
       if (!DestPTy || !SrcPTy)
         return;
 
+      bool InCXXBoundsSafetyMode =
+          Self.LangOpts.CPlusPlus &&
+          Self.LangOpts.isBoundsSafetyAttributeOnlyMode() &&
+          Self.isCXXSafeBuffersEnabledAt(OpRange.getBegin());
+
       /// BoundsSafety: Such case should be handled as BoundsSafetyPointerCast.
       /// \code
       /// int *__bidi_indexable p1;
       /// int *__indexable p2 = (int *__indexable)p1;
       /// \endcode
-      if (Self.getLangOpts().BoundsSafety) {
+      if (Self.getLangOpts().BoundsSafetyAttributes) {
         unsigned DiagKind = 0;
         bool isInvalid = false;
         // The type error may be nested, so any pointer can result in VT errors
```

Comment:
Do we want to assert that `Self.getLangOpts().BoundsSafety || ConvertTy == Sema::Compatible || (ConvertTy == Sema::IncompatibleNonValueTerminatedToValueTerminatedPointer && InCXXBoundsSafetyMode)`? Especially since `CheckValueTerminatedAssignmentConstraints` used to only run under `BoundsSafety`, but now it runs under `BoundsSafetyAttributes` even for C and in C++ without safe buffers mode. We really don't want it to accidentally start returning errors in these modes. Otherwise LGTM

Reply:
Fair point.  But maybe the following is enough?
`Self.getLangOpts().BoundsSafety ||  Self.isCXXSafeBuffersBoundsSafetyInteropEnabledAt(SrcExpr.get()->getBeginLoc())` 

`isCXXSafeBuffersBoundsSafetyInteropEnabledAt` is defined by
```
  bool isCXXSafeBuffersBoundsSafetyInteropEnabledAt(SourceLocation Loc) const {
    return LangOpts.CPlusPlus && LangOpts.isBoundsSafetyAttributeOnlyMode() &&
           isCXXSafeBuffersEnabledAt(Loc);
  }
```

Reply:
That works for me 👍